Nodes format-number and cast-number-to-string do not complile on non-AVR architectures

### Steps to reproduce

- Place `format-number` to a patch
- Connect a number output (e.g. analog-input) to a string input (e.g. text-lcd-16x2)
- Deploy → Upload to Arduino and select Arduino M0

### Expected behavior

The program compiles and uploads OK

### Actual behavior

```
Error occured during uploading: ChildProcessError: Command failed: "/home/nailxx/devel/xod/packages/xod-client-electron/arduino-builders/linux/arduino-builder" -hardware="/home/nailxx/devel/xod/packages/xod-client-electron/arduino-builders/linux/hardware" -hardware="/home/nailxx/.config/xod-client-electron/packages" -libraries="/home/nailxx/devel/xod/packages/xod-client-electron/arduino-libraries" -tools="/home/nailxx/devel/xod/packages/xod-client-electron/arduino-builders/linux/tools" -tools="/home/nailxx/.config/xod-client-electron/packages" -fqbn="arduino:samd:mzero_bl" -build-path="/home/nailxx/.config/xod-client-electron/upload-temp/build" "/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp"
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp: In function 'void xod::xod__core__cast_number_to_string::evaluate(xod::Context)':
/home/nailxx/.config/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1115:27: error: 'dtostrf' was not declared in this scope
     dtostrf(num, 0, 2, str);
```

### How to fix

Currently, the offending nodes use `dtostrf` native function. It is AVR-specific. For other platforms, we should use an stdlib formatting function.

--- a/packages/xod-arduino/platform/runtime.cpp
+++ b/packages/xod-arduino/platform/runtime.cpp
@@ -42,6 +42,66 @@
 #  define pgm_read_ptr(addr) (*(const void **)(addr))
 #endif
 
+//----------------------------------------------------------------------------
+// Compatibilities
+//----------------------------------------------------------------------------
+
+#if !defined(ARDUINO_ARCH_AVR)
+/*
+ * Provide dtostrf function for non-AVR platforms. Although many platforms
+ * provide a stub many others do not. And the stub is based on `sprintf`
+ * which doesn’t work with floating point formatters on some platforms
+ * (e.g. Arduino M0).
+ *
+ * This is an implementation based on `fcvt` standard function. Taken here:
+ * https://forum.arduino.cc/index.php?topic=368720.msg2542614#msg2542614
+ */
+char *dtostrf(double val, int width, unsigned int prec, char *sout) {
+    int decpt, sign, reqd, pad;
+    const char *s, *e;
+    char *p;
+    s = fcvt(val, prec, &decpt, &sign);
+    if (prec == 0 && decpt == 0) {
+        s = (*s < '5') ? "0" : "1";
+        reqd = 1;
+    } else {
+        reqd = strlen(s);
+        if (reqd > decpt) reqd++;
+        if (decpt == 0) reqd++;
+    }
+    if (sign) reqd++;
+    p = sout;
+    e = p + reqd;
+    pad = width - reqd;
+    if (pad > 0) {
+        e += pad;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    if (sign) *p++ = '-';
+    if (decpt <= 0 && prec > 0) {
+        *p++ = '0';
+        *p++ = '.';
+        e++;
+        while ( decpt < 0 ) {
+            decpt++;
+            *p++ = '0';
+        }
+    }
+    while (p < e) {
+        *p++ = *s++;
+        if (p == e) break;
+        if (--decpt == 0) *p++ = '.';
+    }
+    if (width < 0) {
+        pad = (reqd + width) * -1;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    *p = 0;
+    return sout;
+}
+#endif
+
+
 namespace xod {
 //----------------------------------------------------------------------------
 // Type definitions

--- a/packages/xod-client-electron/test-func/fixtures/blink.cpp
+++ b/packages/xod-client-electron/test-func/fixtures/blink.cpp
@@ -681,6 +681,66 @@ template <typename T> class List {
 #  define pgm_read_ptr(addr) (*(const void **)(addr))
 #endif
 
+//----------------------------------------------------------------------------
+// Compatibilities
+//----------------------------------------------------------------------------
+
+#if !defined(ARDUINO_ARCH_AVR)
+/*
+ * Provide dtostrf function for non-AVR platforms. Although many platforms
+ * provide a stub many others do not. And the stub is based on `sprintf`
+ * which doesn’t work with floating point formatters on some platforms
+ * (e.g. Arduino M0).
+ *
+ * This is an implementation based on `fcvt` standard function. Taken here:
+ * https://forum.arduino.cc/index.php?topic=368720.msg2542614#msg2542614
+ */
+char *dtostrf(double val, int width, unsigned int prec, char *sout) {
+    int decpt, sign, reqd, pad;
+    const char *s, *e;
+    char *p;
+    s = fcvt(val, prec, &decpt, &sign);
+    if (prec == 0 && decpt == 0) {
+        s = (*s < '5') ? "0" : "1";
+        reqd = 1;
+    } else {
+        reqd = strlen(s);
+        if (reqd > decpt) reqd++;
+        if (decpt == 0) reqd++;
+    }
+    if (sign) reqd++;
+    p = sout;
+    e = p + reqd;
+    pad = width - reqd;
+    if (pad > 0) {
+        e += pad;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    if (sign) *p++ = '-';
+    if (decpt <= 0 && prec > 0) {
+        *p++ = '0';
+        *p++ = '.';
+        e++;
+        while ( decpt < 0 ) {
+            decpt++;
+            *p++ = '0';
+        }
+    }
+    while (p < e) {
+        *p++ = *s++;
+        if (p == e) break;
+        if (--decpt == 0) *p++ = '.';
+    }
+    if (width < 0) {
+        pad = (reqd + width) * -1;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    *p = 0;
+    return sout;
+}
+#endif
+
+
 namespace xod {
 //----------------------------------------------------------------------------
 // Type definitions

--- a/workspace/blink/__fixtures__/arduino.cpp
+++ b/workspace/blink/__fixtures__/arduino.cpp
@@ -681,6 +681,66 @@ template <typename T> class List {
 #  define pgm_read_ptr(addr) (*(const void **)(addr))
 #endif
 
+//----------------------------------------------------------------------------
+// Compatibilities
+//----------------------------------------------------------------------------
+
+#if !defined(ARDUINO_ARCH_AVR)
+/*
+ * Provide dtostrf function for non-AVR platforms. Although many platforms
+ * provide a stub many others do not. And the stub is based on `sprintf`
+ * which doesn’t work with floating point formatters on some platforms
+ * (e.g. Arduino M0).
+ *
+ * This is an implementation based on `fcvt` standard function. Taken here:
+ * https://forum.arduino.cc/index.php?topic=368720.msg2542614#msg2542614
+ */
+char *dtostrf(double val, int width, unsigned int prec, char *sout) {
+    int decpt, sign, reqd, pad;
+    const char *s, *e;
+    char *p;
+    s = fcvt(val, prec, &decpt, &sign);
+    if (prec == 0 && decpt == 0) {
+        s = (*s < '5') ? "0" : "1";
+        reqd = 1;
+    } else {
+        reqd = strlen(s);
+        if (reqd > decpt) reqd++;
+        if (decpt == 0) reqd++;
+    }
+    if (sign) reqd++;
+    p = sout;
+    e = p + reqd;
+    pad = width - reqd;
+    if (pad > 0) {
+        e += pad;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    if (sign) *p++ = '-';
+    if (decpt <= 0 && prec > 0) {
+        *p++ = '0';
+        *p++ = '.';
+        e++;
+        while ( decpt < 0 ) {
+            decpt++;
+            *p++ = '0';
+        }
+    }
+    while (p < e) {
+        *p++ = *s++;
+        if (p == e) break;
+        if (--decpt == 0) *p++ = '.';
+    }
+    if (width < 0) {
+        pad = (reqd + width) * -1;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    *p = 0;
+    return sout;
+}
+#endif
+
+
 namespace xod {
 //----------------------------------------------------------------------------
 // Type definitions

--- a/workspace/lcd-time/__fixtures__/arduino.cpp
+++ b/workspace/lcd-time/__fixtures__/arduino.cpp
@@ -681,6 +681,66 @@ template <typename T> class List {
 #  define pgm_read_ptr(addr) (*(const void **)(addr))
 #endif
 
+//----------------------------------------------------------------------------
+// Compatibilities
+//----------------------------------------------------------------------------
+
+#if !defined(ARDUINO_ARCH_AVR)
+/*
+ * Provide dtostrf function for non-AVR platforms. Although many platforms
+ * provide a stub many others do not. And the stub is based on `sprintf`
+ * which doesn’t work with floating point formatters on some platforms
+ * (e.g. Arduino M0).
+ *
+ * This is an implementation based on `fcvt` standard function. Taken here:
+ * https://forum.arduino.cc/index.php?topic=368720.msg2542614#msg2542614
+ */
+char *dtostrf(double val, int width, unsigned int prec, char *sout) {
+    int decpt, sign, reqd, pad;
+    const char *s, *e;
+    char *p;
+    s = fcvt(val, prec, &decpt, &sign);
+    if (prec == 0 && decpt == 0) {
+        s = (*s < '5') ? "0" : "1";
+        reqd = 1;
+    } else {
+        reqd = strlen(s);
+        if (reqd > decpt) reqd++;
+        if (decpt == 0) reqd++;
+    }
+    if (sign) reqd++;
+    p = sout;
+    e = p + reqd;
+    pad = width - reqd;
+    if (pad > 0) {
+        e += pad;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    if (sign) *p++ = '-';
+    if (decpt <= 0 && prec > 0) {
+        *p++ = '0';
+        *p++ = '.';
+        e++;
+        while ( decpt < 0 ) {
+            decpt++;
+            *p++ = '0';
+        }
+    }
+    while (p < e) {
+        *p++ = *s++;
+        if (p == e) break;
+        if (--decpt == 0) *p++ = '.';
+    }
+    if (width < 0) {
+        pad = (reqd + width) * -1;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    *p = 0;
+    return sout;
+}
+#endif
+
+
 namespace xod {
 //----------------------------------------------------------------------------
 // Type definitions

--- a/workspace/two-button-switch/__fixtures__/arduino.cpp
+++ b/workspace/two-button-switch/__fixtures__/arduino.cpp
@@ -681,6 +681,66 @@ template <typename T> class List {
 #  define pgm_read_ptr(addr) (*(const void **)(addr))
 #endif
 
+//----------------------------------------------------------------------------
+// Compatibilities
+//----------------------------------------------------------------------------
+
+#if !defined(ARDUINO_ARCH_AVR)
+/*
+ * Provide dtostrf function for non-AVR platforms. Although many platforms
+ * provide a stub many others do not. And the stub is based on `sprintf`
+ * which doesn’t work with floating point formatters on some platforms
+ * (e.g. Arduino M0).
+ *
+ * This is an implementation based on `fcvt` standard function. Taken here:
+ * https://forum.arduino.cc/index.php?topic=368720.msg2542614#msg2542614
+ */
+char *dtostrf(double val, int width, unsigned int prec, char *sout) {
+    int decpt, sign, reqd, pad;
+    const char *s, *e;
+    char *p;
+    s = fcvt(val, prec, &decpt, &sign);
+    if (prec == 0 && decpt == 0) {
+        s = (*s < '5') ? "0" : "1";
+        reqd = 1;
+    } else {
+        reqd = strlen(s);
+        if (reqd > decpt) reqd++;
+        if (decpt == 0) reqd++;
+    }
+    if (sign) reqd++;
+    p = sout;
+    e = p + reqd;
+    pad = width - reqd;
+    if (pad > 0) {
+        e += pad;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    if (sign) *p++ = '-';
+    if (decpt <= 0 && prec > 0) {
+        *p++ = '0';
+        *p++ = '.';
+        e++;
+        while ( decpt < 0 ) {
+            decpt++;
+            *p++ = '0';
+        }
+    }
+    while (p < e) {
+        *p++ = *s++;
+        if (p == e) break;
+        if (--decpt == 0) *p++ = '.';
+    }
+    if (width < 0) {
+        pad = (reqd + width) * -1;
+        while (pad-- > 0) *p++ = ' ';
+    }
+    *p = 0;
+    return sout;
+}
+#endif
+
+
 namespace xod {
 //----------------------------------------------------------------------------
 // Type definitions

